to_msgpack() error: E   TypeError: can not serialize 'ExternalPartition' object

* mashumaro version: 3.15
* Python version:  3.11.8
* Operating System:  MacOS

### Description

A to_msgpack() call that worked in 3.14 no longer works for a particular class in 3.15.


### What I Did

test script:
```
from dataclasses import dataclass, field
from typing import Dict, Any, Optional, Union, List

from mashumaro.mixins.msgpack import DataClassMessagePackMixin

@dataclass
class ExternalPartition(DataClassMessagePackMixin):
    name: str = ""
    description: str = ""
    data_type: str = ""
    meta: Dict[str, Any] = field(default_factory=dict)

@dataclass 
class ExternalTable(DataClassMessagePackMixin):
    location: Optional[str] = None
    file_format: Optional[str] = None
    row_format: Optional[str] = None
    tbl_properties: Optional[str] = None
    partitions: Optional[Union[List[ExternalPartition], List[str]]] = None

def test_partitions_serialization():

    part1 = ExternalPartition(
        name="partition 1",
        description="partition 1",
        data_type="string",
    )

    part2 = ExternalPartition(
        name="partition 2",
        description="partition 2",
        data_type="string",
    )

    ext_table = ExternalTable(
        location="my_location",
        file_format="my file format",
        row_format="row format",
        partitions=[part1, part2],
    )

    ext_table_dict = ext_table.to_dict()
    assert isinstance(ext_table_dict["partitions"][0], dict)

    ext_table_msgpack = ext_table.to_msgpack()
    assert ext_table_msgpack
```
output
```
>       ext_table_msgpack = ext_table.to_msgpack()

tests/functional/artifacts/test_serialization.py:56: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
<string>:12: in __mashumaro_to_msgpack__
    ???
../mashumaro/mashumaro/mixins/msgpack.py:30: in default_encoder
    return msgpack.packb(data, use_bin_type=True)
env/lib/python3.11/site-packages/msgpack/__init__.py:36: in packb
    return Packer(**kwargs).pack(o)
msgpack/_packer.pyx:279: in msgpack._cmsgpack.Packer.pack
    ???
msgpack/_packer.pyx:276: in msgpack._cmsgpack.Packer.pack
    ???
msgpack/_packer.pyx:270: in msgpack._cmsgpack.Packer._pack
    ???
msgpack/_packer.pyx:213: in msgpack._cmsgpack.Packer._pack_inner
    ???
msgpack/_packer.pyx:270: in msgpack._cmsgpack.Packer._pack
    ???
msgpack/_packer.pyx:232: in msgpack._cmsgpack.Packer._pack_inner
    ???
msgpack/_packer.pyx:270: in msgpack._cmsgpack.Packer._pack
    ???
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

>   ???
E   TypeError: can not serialize 'ExternalPartition' object

msgpack/_packer.pyx:257: TypeError
```

msgpack is 1.1.0

--- a/mashumaro/core/meta/types/pack.py
+++ b/mashumaro/core/meta/types/pack.py
@@ -329,7 +329,9 @@ def pack_union(
             spec.copy(type=type_arg, expression="value", owner=spec.type)
         )
         if packer not in packers:
-            if packer == "value":
+            if packer == "value" and not issubclass(
+                get_type_origin(type_arg), Collection
+            ):
                 packers.insert(0, packer)
             else:
                 packers.append(packer)
@@ -356,7 +358,9 @@ def pack_union(
                 )
             else:
                 packer_arg_type_check = f"is {packer_arg_type_names[0]}"
-            if packer == "value":
+            if packer == "value" and not issubclass(
+                packer_arg_type, Collection
+            ):
                 with lines.indent(
                     f"if value.__class__ {packer_arg_type_check}:"
                 ):

--- a/tests/test_union.py
+++ b/tests/test_union.py
@@ -5,8 +5,10 @@ from typing import Any, Dict, List, Union
 
 import pytest
 
-from mashumaro import DataClassDictMixin
+from mashumaro import DataClassDictMixin, pass_through
 from mashumaro.codecs.basic import encode
+from mashumaro.config import BaseConfig
+from mashumaro.dialect import Dialect
 from tests.utils import same_types
 
 
@@ -139,3 +141,27 @@ def test_union_encoding():
             encoded = encode(value, Union[variants])
             assert value == encoded
             assert same_types(value, encoded)
+
+
+def test_union_no_copy_list_with_dataclass_items_or_passed_through_items():
+    class NoCopyListDialect(Dialect):
+        no_copy_collections = (list,)
+
+    @dataclass
+    class Item(DataClassDictMixin):
+        value: int
+
+    @dataclass
+    class Container(DataClassDictMixin):
+        items: Union[list[Item], list[str]]
+
+        class Config(BaseConfig):
+            dialect = NoCopyListDialect
+            serialization_strategy = {str: {"serialize": lambda x: str(x)}}
+
+    item1 = Item(1)
+    item2 = Item(2)
+    items = [item1, item2]
+    container = Container(items=items)
+    data = container.to_dict()
+    assert data == {"items": [{"value": 1}, {"value": 2}]}

